Change mathify to use less memory

karenc opened the pull request:
- Initialize mathjax-node outside of the convertMathML function

  We don't need to reinitialize mathjax-node every time we call the
  convert math function.  This is in preparation of converting batches of
  math instead of all the math at once.

- Change convertMathML to return converted elements and unique css

  convertMathML used to return a list of html, css and svg for each
  element.  The caller then processed the output to get out either the
  html or svg and deduplicate the css.  Instead, just do all of it in
  convertMathML.

- Convert math in batches of 3000 elements

  Instead of converting all the math in the page at once, which builds
  huge arrays, convert 3000 elements at a time to try to reduce memory
  usage.

- Remove MatchWebFonts extension from mathjax config

  MatchWebFonts is designed for pages that have mathematics within text
  that is displayed using web fonts.  It can be configured to check for
  web font arrivals so font size can be determined.
  (http://docs.mathjax.org/en/latest/options/other/MatchWebFonts.html)
  
  The way we use mathjax is we give it an equation to convert, together
  with the font size.  We don't have a page, so this extension isn't
  necessary.
  
  The reason why we should remove it from the mathjax config is it appears
  to have a memory leak the way we're using it.  Specifically, it's these
  lines:
  
  ```
  AJAX.timer.start(AJAX,["checkFonts",this,state.jax[this.id]],
                   CONFIG.fontCheckDelay,CONFIG.fontCheckTimeout);
  ```
  
  https://github.com/mathjax/MathJax/blob/419b0a6eee7eefc0f85e47f7d4f8227ec28b8e57/unpacked/extensions/MatchWebFonts.js#L131-L132
  
  The memory usage goes up with each math element we convert.

- Upgrade puppeteer from 1.5.0 to 1.17.0


- Stop loading resources in puppeteer, uses too much memory

  Add a request interception handler to only load local files with
  extensions `.xhtml`, `.html` and `.css`.
  
  When loading a large book, puppeteer uses way too much memory (>2GB),
  one of the ways to lower the memory usage is to not load the images.

- Use node to replace mathml with converted math

  We were using puppeteer to replace the mathml in the page with svg.
  When serializing the content back out of chrome, we run out of memory
  (for a VM with 2GB ram).
  
  Instead, do this:
  
  1. When looking for math elements in puppeteer / chrome, get the math
     and replace the element with a html comment placeholder.
  2. Convert the math in batches and replace the html comment placeholder
     with the converted math.  At the end of the batch, write the
     converted content so far to the output file.
  3. Read the output file back in and add any css generated by
     mathjax-node.

karenc commented:
Test plan:

1. *(optional)* Clone the cnx-recipes repo:

   ```
   git clone https://github.com/openstax/cnx-recipes.git
   cd cnx-recipes
   ```

1. Check out the `docker-compose-command` branch:

   ```
   git checkout docker-compose-command
   ```

1. Edit `docker-compose.yml` line 67, add `#lower-memory-usage` to the end so that it looks like this:

   ```yaml
     mathify:                                                           
       build: https://github.com/openstax/mathify.git#lower-memory-usage
       image: openstax/mathify
   ```

1. Rebuild the mathify image:

   ```
   docker-compose build mathify
   ```

1. *(optional)* if you don't have a baked book, then do these steps to download physics:

   ```
   docker-compose run --rm fetch-book --with-resources physics
   docker-compose run --rm bake-book physics
   ```

1. Test mathifying physics:

   ```
   docker-compose run --rm mathify-book physics
   ```

Hopefully this doesn't use up all the RAM on your machine and actually creates a `data/physics/collection.mathified.xhtml`.  Have a look at the file in a browser and confirm the math looks ok.

---

The books you probably want to test (because they have a lot of math) are:
 - physics
 - elementary-algebra
 - apphysics
 - prealgebra
 - intermediate-algebra
 - u-physics
 - chemistry

omehes commented:
Tested as followes:
added to docker-compose.yml 'lower-memory-usage'
mathify:                                                           
    build: https://github.com/openstax/mathify.git#lower-memory-usage
    image: openstax/mathify
then
Rebuild the mathify image: docker-compose build mathify``` 

Before these steps I also did git pull to update my cnx-recipes (just in case). Then lowered docker memory size to 2GB and ran mathify on Elementary-Algebra. It didn’t go through. When I did the same at 6GB docker memory, it went OK. It still heavily depends on the docker memory size, it seems. 

Messages:
SVG - Unknown character: U+BA in MathJax_Main,MathJax_Size1,MathJax_AMS
SVG - Unknown character: U+BA in MathJax_Main,MathJax_Size1,MathJax_AMS
14:43:47.470Z  INFO node-typeset: Converted 6000 elements.
(node:9) UnhandledPromiseRejectionWarning: Error: Page crashed!
    at Page._onTargetCrashed (/src/node_modules/puppeteer/lib/Page.js:156:24)

Then:
changed `#lower-memory-usage` to `#wip-memory-usage` in docker-compose.yml
`docker-compose build mathify`

It should now have messages like `memory used: 402` or something like that - OK

cleaned containers
restarted docker
tried on physics and apphysics - mathify didn't work

changed docker memory size to 5GB
physics - OK

philschatz commented on typeset/converter.js:
```diff
@@ -129,73 +146,44 @@ const createMapOfMathMLElements = async (log, inputPath, cssPath, outputPath, ou
     }
-  }, mathSources, PROGRESS_TIME)
-
-  // Inject any CSS that was generated by mathjax-node
```
Does this mean that the HTML+CSS option no longer works?

karenc replied:
If you look under `Injecting MathJax-created CSS...`, you can find the code to inject the css.

philschatz commented on typeset/converter.js:
```diff
@@ -129,73 +146,44 @@ const createMapOfMathMLElements = async (log, inputPath, cssPath, outputPath, ou
-  await writeFile(output, convertedContent)
   log.info(`Content saved. Open "${output}" to see converted file.`)
 
-  await browser.close()
```
Why is the browser no longer being closed?

tomjw64 replied:
@philschatz Still being closed. See https://github.com/openstax/mathify/pull/51/files#diff-f3932dbedd8de4dbca5414e981dddabdR156

karenc replied:
We're not using puppeteer after getting all the math elements so I just closed the browser right after that.

tomjw64 commented on typeset/mjnode.js:
```diff
@@ -1,23 +1,14 @@
 
-const convertMathML = async (log, mathMap/*: Map<string, {xml: string, fontSize: number} */, outputFormat) => {
+const startAPI = async (log) => {
+  mjAPI = require('mathjax-node')
```
I'm interested in the reasoning behind moving this `require` into a function, which is typically something I've heard should be avoided.

karenc replied:
I can't remember why I did that :confused:  I'll move it out.

karenc replied:
Oh I remember now.  If you look in the `convertMathML` function in this file, the first thing it checks is whether `mjAPI` is defined.  I'm going to add a different variable called `mjStarted`, to make this clearer.

tomjw64 commented on typeset/mjnode.js:
```diff
@@ -1,23 +1,14 @@
+let mjAPI;
 
-const convertMathML = async (log, mathMap/*: Map<string, {xml: string, fontSize: number} */, outputFormat) => {
+const startAPI = async (log) => {
```
What is the benefit to this function being async anymore? It doesn't look like any async code or callback is being run in it. Still kind of a noob to async functions. @philschatz might also have an answer for me.

karenc replied:
@tomjw64 I'm very new to this :sweat_smile: so now that you point it out, I don't think we need `async` here.  I'll try removing it.

tomjw64 commented on typeset/converter.js:
```diff
@@ -129,73 +146,44 @@ const createMapOfMathMLElements = async (log, inputPath, cssPath, outputPath, ou
+
+    log.debug(`Inserting converted math elements...`)
+    let convertedContent = []
+    for (const [id, xml] of convertedMathML) {
```
Although `Map` guarantees iteration in the order which the keys were added, since we are adding them in the callback  of a Promise (possibly async) in line 114 of `typeset/mjnode.js`, can we guarantee that these entries are sorted ascending by id? Is it worth sorting `convertedMathML` by id first?

PiotrKozlowski replied:
From what I understand we are passing `mathMap/*: Map<string, {xml: string, fontSize: number} */` to the `convertMathML` and we are getting back `convertedMathML/*: Map<string, {svg || html}> */` where keys in both of this maps are `id`s of math elements, so when we iterate over this map we really doesn't care about order of elements.

karenc replied:
@PiotrKozlowski We care about the order because we're expecting the math elements to be in the order of their positions in the documents.  If they're not in the right order, `indexOf` and `substr` etc won't work correctly.

@tomjw64 You're right, I'm not sure whether that's guaranteed.  I have successfully mathified all the books, but still, so it looks like at least most of the time it should work.  We can't sort the ids because we're only adding ids if the math element doesn't already exist.  May be I should just use the index instead of the math element id. :thinking: I'll change the code and see how it is.

PiotrKozlowski replied:
I meant that both of this maps are using ids of math elements so when you iterate over map like:
and order of id's is `[1, 4, 2, 3]`
then `nextMathNodeIndex = pageContent.indexOf(mathNodePlaceholder(id))` will still find correct node in the page. Isn't it?

tomjw64 replied:
@karenc It is also important to note that when this process fails in it's current state, no error will occur. For example,
```node
'abc'.substring(0,'abc'.indexOf('d')) // => '' (Empty string instead of IndexError or 'ab' like in python)
```
Given this, if a placeholder had been out of order, it seems as if a couple characters would be lost, but the xhtml file would not necessarily be invalidated.This type of silent failure seems like it would be not very fun to debug later. If its too much to ask for sorting, maybe we can throw an error if/when `pageContent.indexOf` returns `-1`

karenc replied:
I've abandoned the html element id and only use the index as the id for the math nodes.  I'm sorting the ids and I added a check for `nextMathNodeIndex === -1`.

karenc replied:
@tomjw64 Thanks for looking at the details!  The `indexOf` test is absolutely necessary.  :+1:

PiotrKozlowski commented on typeset/mjnode.js:
```diff
@@ -92,6 +92,7 @@ const convertMathML = async (log, mathMap/*: Map<string, {xml: string, fontSize:
   const convertedMathMLElements = new Map()
   let prevTime = Date.now()
   let numDone = 0
+  const convertedCss = new Map()
```
Could it be just `Set` since we do not use values?

karenc replied:
I'll try using `Set`, thanks!

PiotrKozlowski commented on typeset/converter.js:
```diff
@@ -131,37 +131,44 @@ const createMapOfMathMLElements = async (log, inputPath, cssPath, outputPath, ou
-        console.error(mathHTML)
-        mathNode.outerHTML = mathHTML
-      }
+  const BATCH_SIZE = 3000
```
Can we allow user to set custom BATCH_SIZE using some kind of flag?

karenc replied:
I'm sure that's possible.

karenc replied:
I've added `--batch-size` or `-b`.  I'm using a regexp to check if it's a valid integer.  Don't know if that's normal :sweat_smile:

PiotrKozlowski commented on typeset/converter.js:
```diff
@@ -38,14 +38,24 @@ const createMapOfMathMLElements = async (log, inputPath, cssPath, outputPath, ou
+  await page.setRequestInterception(true)
+  page.on('request', (interceptedRequest) => {
+    const url = interceptedRequest.url().split('#')[0].split('?')[0]
+    if (url.startsWith('file:') && (url.endsWith('.xhtml') || url.endsWith('.html') || url.endsWith('.css'))) {
```
In which case we need `.xhtml`, `.html` files to be imported?

karenc replied:
That's actually the input file.  If I don't include `.xhtml` (I don't know if the input file can be `.html` :thinking: :question:) then the input file isn't loaded and we can't do anything :sweat_smile:

PiotrKozlowski replied:
Ahh, tricky :D I thought we are launching puppeteer from file, but it was JSDOM functionality.

karenc commented on typeset/converter.js:
```diff
@@ -129,73 +146,44 @@ const createMapOfMathMLElements = async (log, inputPath, cssPath, outputPath, ou
-  }, PROGRESS_TIME)
-
-  const convertedMathML/*: Map<string, {svg, html, css}> */ = await mjnodeConverter.convertMathML(log, new Map(mathEntries), outputFormat)
+  }, PROGRESS_TIME, mathNodePlaceholder('{id}'))
```
Yes, because I couldn't pass `mathNodePlaceholder` as a function to `page.evaluate`, what I decided to do is to create something like `<!-- math node {id} -->` and then inside `page.evaluate`, do `PLACE_HOLDER.replace('{id}', id)`.  So it is still outputting the right thing.  If you have a better solution, I'll be very happy to change it though.

PiotrKozlowski replied:
What about changing this part:
```
// only set an ID if one does not already exist
      if (!mathNode.getAttribute('id')) {
        mathNode.setAttribute('id', `mjnode-${index}`)
        mathNode.classList.add('-remove-id-later')
      }

      const id = mathNode.getAttribute('id')
      // put html comment placeholder for where the converted math should go
      mathNode.outerHTML = PLACE_HOLDER.replace('{id}', id)
```
to something like:
```
const id = mathNode.getAttribute('id') || `mjnode-${index}`
mathNode.outerHTML = PLACE_HOLDER.replace('{id}', id)
```
where `PLACE_HOLDER` is just `const PLACE_HOLDER = "<!-- math node {id} -->"`

karenc replied:
I don't want to have multiple places that defines that constant `<!-- math node ${id} -->`...  I can rename it to `PLACE_HOLDER_TEMPLATE`?

karenc commented:
> Did none of the tests need updating?

@philschatz I never looked at the tests.  Surprisingly they still all pass, but I'll have a look.

PiotrKozlowski commented:
I've tested on physics. It used less than 1GB of RAM. Math looks ok.

karenc commented:
I wrote up the journey to this PR if anyone is interested: https://gist.github.com/karenc/209250357fa0029f43985346cf34187b

omehes commented:
I tried mathify on calculus-vol3 and apphysics books with 1GB docker memory. Works nice and fast.
Also,
tested on various books with docker memory 2GB - all OK

philschatz commented:
One note on relying on `<!-- {id} -->`: Could someone (accidentally) create this comment in their content and cause this code to crash? If so, would it be worth adding a random prefix in the comment to prevent that type of problem?

tomjw64 commented:
> One note on relying on `<!-- {id} -->`: Could someone (accidentally) create this comment in their content and cause this code to crash? If so, would it be worth adding a random prefix in the comment to prevent that type of problem?

I suppose it's possible, but to play devil's advocate here, it is so extremely unlikely I'm not sure it's worth worrying about. Not only would they need to add the comment with a valid id and prefix, they would have to do so in the exact correct order in the content. And at that point, they probably meant to do it.

karenc commented:
If `<!-- math node {id} -->` is a concern, we can possibly generate some random string (uuid?) at the beginning of a session and do `<!-- {uuid} math node {id} -->`, that way it's extra unlikely for someone to accidentally put in the same kind of comment in the html?  Plus it's not hard to implement.

philschatz commented:
It's probably not a concern & therefore not worth worrying about now 😄

karenc commented:
@philschatz  I had a look at the tests.  (I had to because I finally broke them xD)  They're more functional tests and since I haven't really changed the output, they continue to pass.  I had to change the tests because I added an extra argument `batchSize` to the `createMapOfMathMLElements` function.

@tomjw64 @philschatz @PiotrKozlowski I've finished addressing all the comments.  Could you have a look and if they're ok, I'll squash the `fixup!` commits and remove `WIP`.

PiotrKozlowski commented on typeset/start.js:
```diff
@@ -28,13 +28,22 @@ const argv = yargs
 const pathToInput = path.resolve(argv.xhtml)
 const pathToCss = argv.css ? path.resolve(argv.css) : null
 let outputFormat = 'html'
+let batchSize = Number.isInteger(argv.batchSize) ? argv.batchSize : 3000
```
You can use shorter syntax:
`let batchSize = Number(argv.batchSize) || 3000`

karenc replied:
Done :+1:

tomjw64 commented on typeset/converter.js:
```diff
@@ -94,26 +109,21 @@ const createMapOfMathMLElements = async (log, inputPath, cssPath, outputPath, ou
-
-      const id = mathNode.getAttribute('id')
+      // put html comment placeholder for where the converted math should go
+      mathNode.outerHTML = PLACE_HOLDER_TEMPLATE.replace('{id}', index)
```
Since we are iterating though `mathNodes` and it is constructed via `[...mathMLNodes, ...latexNodes]`, it is not guaranteed to be in the markups DFS order. This has implications in serialization since a latex node could appear _before_ a mathml node yet have an id _greater_ than the mathml node. If we assume that latex and mathml are never intermixed in our content, this is fine. This could probably be fixed by constructing `mathNodes` with a `querySelector` that covers both mathML and latex if possible.
I'll approve since there is now there error checking when an id cannot be found, but we should be wary of how mathify will handle content mixing MathML and LaTeX going forward.

karenc replied:
@tomjw64 This is now fixed in the latest commit.
